module/zfs/zfs_ioctl.c

```diff
@@ -5843,6 +5903,9 @@ zfs_ioctl_init(void)
 	zfs_ioctl_register_dataset_nolog(ZFS_IOC_USERSPACE_UPGRADE,
 	    zfs_ioc_userspace_upgrade, zfs_secpolicy_userspace_upgrade,
 	    POOL_CHECK_SUSPENDED | POOL_CHECK_READONLY);
+	zfs_ioctl_register_dataset_nolog(ZFS_IOC_USEROBJSPACE_UPGRADE,
+	    zfs_ioc_userobjspace_upgrade, zfs_secpolicy_userspace_upgrade,
+	    POOL_CHECK_SUSPENDED | POOL_CHECK_READONLY);
```

Comment:
the objset may not be owned at the time of ```dmu_objset_sync()``` therefore it will indeed meet the issue you mentioned before, which is the dataset could be released at the time of upgrade task being running.

Of course this issue can be easily fixed by holding this dataset in ```dmu_objset_sync()``` and add a flag to mark that this dataset is being held by upgrade task therefore it can be released at the end of upgrade thread. But I don't think it's necessary by adding this extra complexity.